hooks/mari/publish_mipmaps.py

```diff
@@ -91,82 +108,195 @@ def validate(self, task_settings, item):
-            mipmap_paths = self.create_mipmaps_for_seq(src_files, target_path)
+            try:
+                # ensure the publish folder exists:
+                publish_folder = os.path.dirname(target_path)
+                ensure_folder_exists(publish_folder)
 
-        except Exception as e:
-            raise TankError("Failed to publish file for item '%s': %s" % (item.name, str(e)))
+                # Find source file from current item.properties.sg_publish_data_list
+                # since product of publish_texture is the source file, and the item
+                # is shared between the plugins
+                source_path = item.properties["sg_publish_data_list"][0]["path"]["local_path"]
 
-        self.logger.debug(
-            "Published file for item '%s' to '%s'." % (item.name, target_path)
-        )
+                # get all associated files if this is a file sequence
+                seq_path = publisher.util.get_frame_sequence_path(source_path)
+                if seq_path:
+                    src_files = publisher.util.get_sequence_path_files(source_path)
+                else:
+                    src_files = [source_path]
 
-        return mipmap_paths
+                # write mipmaps for each file in the sequence
+                item.properties["mipmap_paths"].extend(self.create_mipmaps_for_seq(src_files, target_path))
+
+            except Exception as e:
+                raise TankError("Failed to publish file for item '%s': %s" % (item.name, str(e)))
+
+            self.logger.debug(
+                "Published file for item '%s' to '%s'." % (item.name, target_path)
+            )
+
+        return item.properties["mipmap_paths"]
+
+    def publish(self, task_settings, item):
+        """
+        Executes the publish logic for the given item and task_settings.
+
+        :param task_settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the task_settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+        # TODO: check if Publish Textures plugin was successfully run for the same item?
+        publisher = self.parent
+
+        # Get item properties populated by validate method
+        publish_type          = item.properties.publish_type
+        publish_version       = item.properties.publish_version
+        publish_user          = item.get_property("publish_user", default_value=None)
+
+        # handle publishing of files first
+        self.publish_files(task_settings, item)
+
+        # Get any upstream dependency paths
+        dependency_paths = item.properties.get("publish_dependency_paths", [])
+
+        # Find upstream dependency from current item.properties.sg_publish_data_list
+        # since product of publish_texture is the upstream dependency, and the item
+        # is shared between the plugins
+        dependency_ids = [x["id"] for x in item.properties.get("sg_publish_data_list")]
```

Comment:
Same as getting source file. item sharing feels icky.

Reply:
This functionality is already generalized in publish_session.py with the `self._get_dependency_paths()` and `self._get_dependency_ids()` methods. Perhaps we should move this up the inheritance chain to publish.py instead?